api/agent/agent.go

```diff
@@ -266,88 +266,114 @@ func (a *agent) getSlot(ctx context.Context, call *call) (Slot, error) {
 
 	isHot := protocol.IsStreamable(protocol.Protocol(call.Format))
 	if isHot {
-		// For hot requests, we use a long lived slot queue, which we use to manage hot containers
-		call.slots = a.slotMgr.getHotSlotQueue(call)
 		start := time.Now()
 
-		call.slots.enterState(SlotQueueWaiter)
-		s, err := a.launchHot(ctx, call)
-		call.slots.exitStateWithLatency(SlotQueueWaiter, uint64(time.Now().Sub(start).Seconds()*1000))
+		// For hot requests, we use a long lived slot queue, which we use to manage hot containers
+		var isNew bool
+		call.slots, isNew = a.slotMgr.getSlotQueue(call)
+		if isNew {
+			go a.hotLauncher(ctx, call)
+		}
 
+		s, err := a.waitHot(ctx, call)
+		call.slots.exitStateWithLatency(SlotQueueWaiter, uint64(time.Now().Sub(start).Seconds()*1000))
 		return s, err
 	}
 
 	return a.launchCold(ctx, call)
 }
 
-// launchHot checks with slot queue to see if a new container needs to be launched and waits
-// for available slots in the queue for hot request execution.
-func (a *agent) launchHot(ctx context.Context, call *call) (Slot, error) {
+// hotLauncher is spawned in a go routine for each slot queue to monitor stats and launch hot
+// containers if needed. Upon shutdown or activity timeout, hotLauncher exits and during exit,
+// it destroys the slot queue.
+func (a *agent) hotLauncher(ctx context.Context, call *call) {
+
+	// Let use 60 minutes or 2 * IdleTimeout as hot queue idle timeout, pick
+	// whichever is longer. If in this time, there's no activity, then
+	// we destroy the hot queue.
+	timeout := time.Duration(60) * time.Minute
+	idleTimeout := time.Duration(call.IdleTimeout) * time.Second * 2
+	if timeout < idleTimeout {
+		timeout = idleTimeout
+	}
 
+	logger := common.Logger(ctx)
+	logger.WithField("launcher_timeout", timeout).Info("Hot function launcher starting")
 	isAsync := call.Type == models.TypeAsync
 
-launchLoop:
+	defer call.slots.destroySlotQueue()
```

Comment:
this works because the following loop invariant is(?):

* server shutdown (need to pin down?)
* destroySlotQueue will have removed this slot queue from the slot queue bucket, 

correct? we could just let the reference die but we also want to signal anyone waiting on the cond ? is that possible since we've confirmed that this channel is idle and removed the ability to get a reference to it [safely]?

for server shutdown, we want to wait for all tasks to finish and for all containers to shut down. i think at present we're in a weird situation where we can have calls come in, they wait for a slot, but the server got shutdown so no containers are going to launch and no slots are going to be handed out, so we just wait for the calls to timeout. seems like maybe in waitHot we should check server shutdown and handle this with an immediate timeout (as opposed to waiting, potentially 60 minutes for async tasks). we can punt on fixing the shutdown in this patch, if the above is true we should probably try to close it though.

Reply:
1) in waitHot() we should check shutdown to avoid that timeout issue
2) I think we can get rid of destroySlotQueue and related close logic in it, with idle check, this is now dead code, nobody should be waiting.

Reply:
With shutdown() case, destroySlotQueue() does wake up other waiters, but I need to add 'close' channel in startDequeuer() again to make this work properly. With that, waitHot() will wake up and find that the channel is now closed. Which addresses the shutdown/timeout issue....

Reply:
So, I have two options I think:
1) wait for shutdown in waitHot and try to simplify destroySlotQueue since now we might need to wake up waiters.
2) add closing of channel in startDequeuer() to let waitHot() detect the shutdown case.

'1' seems better, which I'm looking into...